WordPress/gutenberg
# Block API: Add Block Context support

Closes #19685
Closes #4671
Previously: #19572
WordPress Trac ticket: https://core.trac.wordpress.org/ticket/49927.

This pull request seeks to implement the means for a block to provide and consume contextual values. In many ways, it is based on the earlier work of #19572, but there are some key differences in the specifics of how context is defined. I've included a fair bit of documentation here as well, which was useful for me in validating (["rubber duck"-ing](https://en.wikipedia.org/wiki/Rubber_duck_debugging)) the underlying implementation. This documentation should prove useful for understanding what it is, how it works, and why it's proposed to be implemented the way it is.

Differences from #19572:

- Providing context occurs separately from the block's `attributes` schema, as its own `providesContext` property.
- There's no association from the consumer of where a context value is assumed to be defined. Instead, a provider grants a value via a key corresponding to the name of one of its own attributes, and a consumer references it by name.
   - For more information, granted downsides, and possible future enhancements, refer to the comment at https://github.com/WordPress/gutenberg/issues/19685#issuecomment-603848589 .
- ~Server-side, the implementation proposed here does not modify the function signatures of either `render_block` or a block's `render_callback`. Instead, both the block context of the current hierarchy and the block itself are assigned as globals which can be referenced in the `render_callback`.~ In later revisions of this pull request, an enhancement was achieved where the full `$block` instance is passed as an argument to `render_callback`, retaining backward compatibility by implementing `ArrayAccess` where the instance's array members are short-hand to its own attributes. See https://github.com/WordPress/gutenberg/pull/21467#issuecomment-613635022 for more information.
   - [See documentation](https://github.com/WordPress/gutenberg/blob/c6ea48219de1ffacd0016226051691ce3828cb57/docs/designers-developers/developers/block-api/block-context.md#using-block-context)
   - For more information and granted downsides, refer to the comment at https://github.com/WordPress/gutenberg/issues/19685#issuecomment-603867889 .
   - ~This is future-compatible with revisions to the function signatures, if there's a desire to go this route. I did not feel that this decision should block the implementation. I've remarked to the future possibility of revised function signatures in the above-referenced documentation.~ Later changes of the pull request implemented said future-compatible revisions.
   - ~This also solves requirements where some developers have sought to access other properties of `$block` while within `render_callback`, avoiding needing to resort to workarounds using `render_block` hook (e.g. #19991).~ Later changes of the pull request serve as a solution to this need.
- Block context is assigned from a provider only from the point of `InnerBlocks`, as an optimization to avoid establishing a context unnecessarily if there are no children to consume from it.
- It does not implement any new blocks. For demonstration purposes, the existing Post Title block has been updated to use block context. This was also the case in #19572, but currently it is only a static representation of the title, and does not yet support editing.

**Open Questions:**

- "Default" context: Is it reasonable to establish the default "post" context? There is certainly prior art for this, given that `render_block` already manipulates and assigns the `$post` global. I'd worry if this might lead to a situation where there's not clear expectations around what should or shouldn't be available by default. "Post" is a particularly special case and may be fine to assume it would be unique in this regard.
- Does context need to be available to `save` ? I've assumed not, given that this may lead to unintended invalidations, and that at least for the current set of use-cases, context would be evaluated server-side in a `render_callback`.
- The comment at https://github.com/WordPress/gutenberg/issues/19685#issuecomment-603848589 left open the question about whether these keys ought to be namespaced to avoid conflict. While this is still a possibility and the proposed implementation may be overly simplistic (i.e. naive), it may also be fair to say that context will be reserved for very specific usages, and not widely applied enough so as to have high risk for conflict.
- Minor: Does it make sense that a `context` object always be provided, even when the block does not define any context to consume? My decision to pass an empty object is largely for developer experience, anticipating potential errors if a developer were to assume that the object is present and attempt unguarded access to properties on that (possibly-undefined) value.

**In Progress:**

- ~I am observing PHP warning notices in the "Site Editor" page related to these changes, which I've yet to resolve. They appear to be related to the assignment of the `$block` global in the overridden `render_block` implementation.~ Fixed, see https://github.com/WordPress/gutenberg/pull/21467#issuecomment-612944434.

**Testing Instructions:**

Given that this pull request only ports the Post Title experimental full-site editing block, testing can be a bit cumbersome at the moment, especially if you've not yet defined templates for full-site editing mode.

My recommendation for testing is as follows:

1. Navigate to Gutenberg > Experiments
2. Enable "Full Site Editing" and click "Save Changes"
3. Navigate to Posts > Add New
4. Insert a Post Title block
5. Observe that the block inherits the title of the post and updates if you change the title
6. Preview the post in a separate tab
   - Note: If you don't have templates on your site, you may see a message "No template found". This is expected, and can be disregarded for now.
7. Navigate to Gutenberg > Experiments
8. Disable "Full Site Editing" and click "Save Changes"
9. In your separate preview tab, refresh the page.
10. Observe that the Post Title block displays the title of the current post
